axiom/nr_txn.c

```diff
@@ -3293,6 +3296,91 @@ bool nr_txn_log_decorating_enabled(nrtxn_t* txn) {
   return false;
 }
 
+#define ENSURE_LOG_LEVEL_NAME(level_name) \
+  (NULL == level_name ? "UNKNOWN" : level_name)
+
+static void log_event_set_linking_metadata(nr_log_event_t* e,
+                                           nrtxn_t* txn,
+                                           nrapp_t* app) {
+  char* trace_id = NULL;
+  char* span_id = NULL;
+
+  if (NULL == e) {
+    return;
+  }
+
+  if (nrlikely(txn)) {
+    trace_id = nr_txn_get_current_trace_id(txn);
+    nr_log_event_set_trace_id(e, trace_id);
+    nr_free(trace_id);
+
+    span_id = nr_txn_get_current_span_id(txn);
+    nr_log_event_set_span_id(e, span_id);
+    nr_free(span_id);
+
+    nr_log_event_set_entity_name(e, txn->primary_app_name);
+  }
+
+  if (nrlikely(app)) {
+    nr_log_event_set_hostname(e, nr_app_get_host_name(app));
+    nr_log_event_set_guid(e, nr_app_get_entity_guid(app));
+  }
+}
+
+static nr_log_event_t* log_event_create(const char* log_level_name,
+                                        const char* log_message,
+                                        nrtime_t timestamp,
+                                        nrtxn_t* txn,
+                                        nrapp_t* app) {
+  nr_log_event_t* e = nr_log_event_create();
+  if (NULL == e) {
+    return e;
+  }
+  nr_log_event_set_log_level(e, ENSURE_LOG_LEVEL_NAME(log_level_name));
+  nr_log_event_set_message(e, log_message);
+  nr_log_event_set_timestamp(e, timestamp);
+
+  log_event_set_linking_metadata(e, txn, app);
+
+  return e;
+}
+
+static void nr_txn_add_log_event(nrtxn_t* txn,
+                                 const char* log_level_name,
+                                 const char* log_message,
+                                 nrtime_t timestamp,
+                                 nrapp_t* app) {
+  nr_random_t* rnd = NULL;
+  nr_log_event_t* e = NULL;
+  bool event_dropped = false;
+
```

Comment:
check txn, log_level_name, log_message, app for null before going on?
Even if it is currently called from a function that already checks those, we want to future proof it in case it is called from another function in the future.  it can be an nrunlikely wrap, but should be checked just in case.

Reply:
`nrunlikely(NULL == txn)` checks added in 00da593846c2ab78c36c271da3b5a21225fac560. `log_event_set_linking_metadata` already checked if `NULL != app` and there's no need to abort if `app == NULL` earlier.